rosidl_typesupport_connext_cpp/resource/msg__type_support.cpp.em

```diff
@@ -233,99 +193,113 @@ convert_dds_message_to_ros(
-  bool ignore_local_publications,
-  void * untyped_ros_message,
-  bool * taken,
-  void * sending_publication_handle)
+to_cdr_stream__@(spec.base_type.type)(
+  const void * untyped_ros_message,
+  ConnextStaticCDRStream * cdr_stream)
 {
-  if (!untyped_topic_reader) {
-    throw std::runtime_error("topic reader handle is null");
+  if (!cdr_stream) {
+    return false;
   }
   if (!untyped_ros_message) {
-    throw std::runtime_error("ros message handle is null");
-  }
-  if (!taken) {
-    throw std::runtime_error("taken handle is null");
+    return false;
   }
 
-  DDSDataReader * topic_reader = static_cast<DDSDataReader *>(untyped_topic_reader);
+  // cast the untyped to the known ros message
+  const @(spec.base_type.pkg_name)::@(subfolder)::@(spec.base_type.type) & ros_message =
+    *(const @(spec.base_type.pkg_name)::@(subfolder)::@(spec.base_type.type) *)untyped_ros_message;
 
-  @(spec.base_type.pkg_name)::@(subfolder)::dds_::@(spec.base_type.type)_DataReader * data_reader =
-    @(spec.base_type.pkg_name)::@(subfolder)::dds_::@(spec.base_type.type)_DataReader::narrow(topic_reader);
-  if (!data_reader) {
-    fprintf(stderr, "failed to narrow data reader\n");
+  // create a respective connext dds type
+  @(spec.base_type.pkg_name)::@(subfolder)::dds_::@(spec.base_type.type)_ * dds_message = @(spec.base_type.pkg_name)::@(subfolder)::dds_::@(spec.base_type.type)_TypeSupport::create_data();
+  if (!dds_message) {
     return false;
   }
 
-  @(spec.base_type.pkg_name)::@(subfolder)::dds_::@(spec.base_type.type)_Seq dds_messages;
-  DDS_SampleInfoSeq sample_infos;
-  DDS_ReturnCode_t status = data_reader->take(
-    dds_messages,
-    sample_infos,
-    1,
-    DDS_ANY_SAMPLE_STATE,
-    DDS_ANY_VIEW_STATE,
-    DDS_ANY_INSTANCE_STATE);
-  if (status == DDS_RETCODE_NO_DATA) {
-    *taken = false;
-    return true;
+  // convert ros to dds
+  if (!convert_ros_message_to_dds(ros_message, *dds_message)) {
+    return false;
   }
-  if (status != DDS_RETCODE_OK) {
-    fprintf(stderr, "take failed with status = %d\n", status);
+
+  // call the serialize function for the first time to get the expected length of the message
+  unsigned int expected_length;
+  if (@(spec.base_type.type)_Plugin_serialize_to_cdr_buffer(
+      NULL,
+      &expected_length,
+      dds_message) != RTI_TRUE)
+  {
+    fprintf(stderr, "failed to call @(spec.base_type.type)_Plugin_serialize_to_cdr_buffer()\n");
     return false;
   }
+  cdr_stream->buffer_length = expected_length;
+  if (cdr_stream->buffer_length > (std::numeric_limits<unsigned int>::max)()) {
+    fprintf(stderr, "cdr_stream->buffer_length, unexpectedly larger than max unsigned int\n");
+    return false;
+  }
+  if (cdr_stream->buffer_capacity < cdr_stream->buffer_length) {
+    cdr_stream->allocator.deallocate(cdr_stream->buffer, cdr_stream->allocator.state);
+    cdr_stream->buffer = static_cast<char *>(cdr_stream->allocator.allocate(cdr_stream->buffer_length, cdr_stream->allocator.state));
```

Comment:
After allocating a bigger buffer the `capacity` of `cdr_stream` is never updated. See http://build.ros2.org/view/Rci/job/Rci__nightly-connext_ubuntu_focal_amd64/29/testReport/rclcpp/TestSerializedMessage/serialization/ for a test failing for a while.

See ros2/rosidl_typesupport_connext#58 for the proposed fix.